src/vm_pages.rs

```diff
@@ -315,6 +315,58 @@ impl VmRegionList {
         Ok(())
     }
 
+    // Removes the region at [`start`, `end`) of type `region_type`.
+    fn remove(
+        &mut self,
+        start: GuestPageAddr,
+        end: GuestPageAddr,
+        region_type: VmRegionType,
+    ) -> Result<()> {
+        let mut index = self
+            .regions
+            .iter()
+            .enumerate()
+            .find(|(_, r)| r.start <= start && r.end >= end && r.region_type == region_type)
+            .map(|(i, _)| i)
+            .ok_or(Error::InvalidMapRegion)?;
+
+        // Unwrap ok here since we know 'index' is valid.
+        let region = self.regions.get(index).unwrap();
+        let r_start = region.start;
+        let r_end = region.end;
+
+        // Remove the memory region.
+        self.regions.remove(index);
+
+        // Add the remaining range at the start as a new region.
+        if r_start < start {
+            let new_region = VmRegion {
+                start: r_start,
+                end: start,
+                region_type,
+            };
+            self.regions
+                .try_insert(index, new_region)
+                .map_err(|_| Error::InsufficientVmRegionSpace)?;
```

Comment:
Need to check that we have space for the new region before we go to remove the old one, otherwise we could leave the list in an inconsistent state.

Reply:
Nice catch. Thanks.
Fixed.